### PR TITLE
[C99] Claim conformance for _Complex support

### PR DESCRIPTION
**Pull request by AaronBallman**
There's so much overlap between the cited papers so this condenses the status page into a single entry rather than trying to test conformance against multiple papers doing conflicting things.

### CODE REVIEW

**Comment by AaronBallman** [comment]
I'm not convinced I have sufficient tests, so I'm especially interested in some help figuring out what else to test and how to go about it. I don't believe we need to test library behavior here because Clang does not provide any library support for complex types in freestanding mode; we expect the user to link against a C standard library that has such support.

**Comment by AaronBallman** [comment]
> > I'm not convinced I have sufficient tests, so I'm especially interested in some help figuring out what else to test and how to go about it. I don't believe we need to test library behavior here because Clang does not provide any library support for complex types in freestanding mode; we expect the user to link against a C standard library that has such support.
> 
> I haven't looked at it in details yet, but the test needs to run on Linux and Windows.

I don't think there's anything platform-specific about our support at this level, so the test has no triples in the RUN line (so it will be run on all triples we test on which includes Linux and Windows). Or do you mean something else?

**Comment by zahiraam** [comment]
> > > I'm not convinced I have sufficient tests, so I'm especially interested in some help figuring out what else to test and how to go about it. I don't believe we need to test library behavior here because Clang does not provide any library support for complex types in freestanding mode; we expect the user to link against a C standard library that has such support.
> > 
> > 
> > I haven't looked at it in details yet, but the test needs to run on Linux and Windows.
> 
> I don't think there's anything platform-specific about our support at this level, so the test has no triples in the RUN line (so it will be run on all triples we test on which includes Linux and Windows). Or do you mean something else?

I guess at the IR level no. It's just that I have recently found that this test:
`#include <stdio.h>`
`#include <complex.h>`

`int main()`
`{`
    `float complex b = -8.45345913e+29f + -1.23588801e+31f * I;`
   ` float complex a = 1.25284138e+31f + -7.91789592e+30f * I;`
    `float complex c = a / b;`
    `printf("%f %f\n", creal(c), cimag(c));`
    `return 0;`
`}`
But fails on Windows.

**Comment by frederick-vs-ja** [comment]
> But fails on Windows.

This seems because of that MS UCRT's complex math functions are non-conforming.

I _guess_ they could work if we write the following before including UCRT's `<complex.h>`.
```C
#define _C_COMPLEX_T
typedef double _Complex _C_double_complex;
typedef float _Complex _C_float_complex;
typedef long double _Complex _C_ldouble_complex;
```

**Comment by philnik777** [comment]
Note that
```c++
auto div(_Complex float lhs, _Complex float rhs) {
  return lhs / rhs;
}

int main() {
  return __real div(1.f, 2.f);
}
```
will fail to link on windows due to `__divsc3` not being available. Similar programs probably also fail due to other compiler-rt functions not being available.

**Comment by AaronBallman** [comment]
> Note that
> 
> ```c++
> auto div(_Complex float lhs, _Complex float rhs) {
>   return lhs / rhs;
> }
> 
> int main() {
>   return __real div(1.f, 2.f);
> }
> ```
> 
> will fail to link on windows due to `__divsc3` not being available. Similar programs probably also fail due to other compiler-rt functions not being available.

I *think* Clang still gets to claim conformance here because we handle the language side of things correctly and the user is responsible for linking to a runtime library with appropriate support for the platform. However, this sure does straddle the line in some ways, so I could see this being a reason to claim "partial" conformance.

**Review by jcranmer-intel — COMMENTED** [review]
You're missing checks for type domain rules, so things like:

- converting between `float _Complex` and `double _Complex`
- common type of `float _Complex` and `double`
- result of `int` and `float _Complex`
- complex types not allowed in increment/decrement operators
- complex types permitted in unary +, -

Not sure how to test this, but _Complex float not promoting to _Complex double like float does to double is also worth slotting under complex conformance. Also 6.7.8p10 (static-storage duration arithmetic types (which include complex types) are initialized to 0).

File: clang/test/C/C99/n809.c [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+double _Imaginary di1; // expected-error {{imaginary types are not supported}}
+_Imaginary double di2; // expected-error {{imaginary types are not supported}}
+
+long double _Imaginary ldi1; // expected-error {{imaginary types are not supported}}
+_Imaginary long double ldi2; // expected-error {{imaginary types are not supported}}
+
+// Each complex type has the same representation and alignment as an array
+// containing two elements of the corresponding real type.
+_Static_assert(sizeof(float _Complex) == sizeof(struct { float mem[2]; }), "");
+_Static_assert(_Alignof(float _Complex) == _Alignof(struct { float mem[2]; }), "");
```
[the remark this answers is not in the document]
Comment (AaronBallman) [reply clang/test/C/C99/n809.c:39]: I think this is a property we want to guarantee, so it seems reasonable-ish to test it here as part of our conformance testing.

File: clang/test/C/C99/n809.c [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+_Static_assert(sizeof(double _Complex) == sizeof(struct { double mem[2]; }), "");
+_Static_assert(_Alignof(double _Complex) == _Alignof(struct { double mem[2]; }), "");
+
+_Static_assert(sizeof(long double _Complex) == sizeof(struct { long double mem[2]; }), "");
+_Static_assert(_Alignof(long double _Complex) == _Alignof(struct { long double mem[2]; }), "");
+
+// The first element corresponds to the real part and the second element
+// corresponds to the imaginary part.
+_Static_assert(__real((float _Complex){ 1.0f, 2.0f }) == 1.0f, "");
+_Static_assert(__imag((float _Complex){ 1.0f, 2.0f }) == 2.0f, "");
```
[the remark this answers is not in the document]
Comment (AaronBallman) [reply clang/test/C/C99/n809.c:50]: The other alternative would be to use a builtin function, I can go either way. Maybe I should use one of each to show they're equivalent?

File: clang/www/c_status.html [hunk 3]
```diff
@@ -373,6 +355,10 @@ <h2 id="c99">C99 implementation status</h2>
       <td class="full" align="center">Yes</td>
     </tr>
 </table>
+<span id="complex">(2): Clang supports <code>_Complex</code> type specifiers but
+does not support <code>_Imaginary</code> type specifiers. Support for
+<code>_Imaginary</code> is optional in C99 which is why Clang is fully conforming.
```
[the remark this answers is not in the document]
Comment (AaronBallman) [reply clang/www/c_status.html:360]: Done

**Comment by AaronBallman** [comment]
> You're missing checks for type domain rules, so things like:
> 
>     * converting between `float _Complex` and `double _Complex`
> 
>     * common type of `float _Complex` and `double`
> 
>     * result of `int` and `float _Complex`
> 
>     * complex types not allowed in increment/decrement operators
> 
>     * complex types permitted in unary +, -

Thanks for the suggestions, these are now being tested.

> Not sure how to test this, but _Complex float not promoting to _Complex double like float does to double is also worth slotting under complex conformance. Also 6.7.8p10 (static-storage duration arithmetic types (which include complex types) are initialized to 0).

I'll add tests for these as well.

**Comment by philnik777** [comment]
> > Note that
> > ```c++
> > auto div(_Complex float lhs, _Complex float rhs) {
> >   return lhs / rhs;
> > }
> > 
> > int main() {
> >   return __real div(1.f, 2.f);
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > will fail to link on windows due to `__divsc3` not being available. Similar programs probably also fail due to other compiler-rt functions not being available.
> 
> I _think_ Clang still gets to claim conformance here because we handle the language side of things correctly and the user is responsible for linking to a runtime library with appropriate support for the platform. However, this sure does straddle the line in some ways, so I could see this being a reason to claim "partial" conformance.

IIRC the problem is that you can't get compiler-rt on windows. If it was just a matter of compiling compiler-rt with clang I'd be 100% with you. Partial conformance with a note that some operations require compiler-rt functions that aren't available on windows seems like a good compromise to me.

**Comment by AaronBallman** [comment]
> > > Note that
> > > ```c++
> > > auto div(_Complex float lhs, _Complex float rhs) {
> > >   return lhs / rhs;
> > > }
> > > 
> > > int main() {
> > >   return __real div(1.f, 2.f);
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > will fail to link on windows due to `__divsc3` not being available. Similar programs probably also fail due to other compiler-rt functions not being available.
> > 
> > 
> > I _think_ Clang still gets to claim conformance here because we handle the language side of things correctly and the user is responsible for linking to a runtime library with appropriate support for the platform. However, this sure does straddle the line in some ways, so I could see this being a reason to claim "partial" conformance.
> 
> IIRC the problem is that you can't get compiler-rt on windows. If it was just a matter of compiling compiler-rt with clang I'd be 100% with you. Partial conformance with a note that some operations require compiler-rt functions that aren't available on windows seems like a good compromise to me.

Yeah, that seems like a reasonable compromise. I'll make updates for that, thanks!

**Review by philnik777 — APPROVED** [review]
LGTM from my side. I'll leave the conformance test details to the folks who know what they're doing.

**Review by zahiraam — APPROVED** [review]

**Comment by zahiraam** [comment]
LGTM. Thanks.

**Review by jcranmer-intel — APPROVED** [review]

**Comment by h-vetinari** [comment]
I'm surprised about the "compiler-rt is not supported on windows" comment - in our distribution, we've been building compiler-rt on windows for years, and I don't remember significant issues.

 It's true that `__divsc3` is not available _without_ compiler-rt (i.e. just relying on the msvc runtime libraries), but together with compiler-rt these things are generally possible? Not that a similar [situation](https://github.com/llvm/llvm-project/issues/54596) exists on `osx-arm` for example - in certain situations, linking `compiler-rt` is unavoidable.

**Comment by AaronBallman** [comment]
> I'm surprised about the "compiler-rt is not supported on windows" comment - in our distribution, we've been building compiler-rt on windows for years, and I don't remember significant issues.
> 
> It's true that `__divsc3` is not available _without_ compiler-rt (i.e. just relying on the msvc runtime libraries), but together with compiler-rt these things are generally possible? Note that a similar [situation](https://github.com/llvm/llvm-project/issues/54596) exists on `osx-arm` for example - in certain situations, linking `compiler-rt` is unavoidable.

CC @petrhosek as compiler-rt CRT code owner -- is compiler-rt officially supported for Windows? The official documentation does not mention it under platform support: https://compiler-rt.llvm.org/

**Pull request merged**